File: global-ip-frontend/src/services/roleRequestService.ts
import api from './api';

export interface RoleRequest {
  id: string;
  userId: string;
  requestedRole: string;
  status: 'PENDING' | 'APPROVED' | 'REJECTED' | 'WAITLISTED';
  requestedAt: string;
  reviewedBy?: string;
  reviewedAt?: string;
}

export interface RoleRequestAdminView {
  requestId: number;
  userId: number;
  username: string;
  email: string;
  requestedRole: string;
  requestedAt: string;
}

export interface ApiResponse {
  message: string;
}

class RoleRequestService {
  /**
   * USER: Request admin role
   */
  async requestAdminRole(): Promise<ApiResponse> {
    const response = await api.post<ApiResponse>('/role-requests/admin');
    return response.data;
  }

  /**
   * ADMIN: Get all pending role requests
   */
  async getPendingRequests(): Promise<RoleRequestAdminView[]> {
    const response = await api.get<RoleRequestAdminView[]>('/role-requests/pending');
    return response.data;
  }

  /**
   * ADMIN: Approve a role request
   */
  async approveRequest(requestId: number): Promise<ApiResponse> {
    const response = await api.post<ApiResponse>(`/role-requests/adminOnly/${requestId}/approve`);
    return response.data;
  }

  /**
   * ADMIN: Reject a role request
   */
  async rejectRequest(requestId: number): Promise<ApiResponse> {
    const response = await api.post<ApiResponse>(`/role-requests/adminOnly/${requestId}/reject`);
    return response.data;
  }

  /**
   * ADMIN: Waitlist a role request
   */
  async waitlistRequest(requestId: number): Promise<ApiResponse> {
    const response = await api.post<ApiResponse>(`/role-requests/adminOnly/${requestId}/waitlist`);
    return response.data;
  }
}

export default new RoleRequestService();

File: global-ip-core/src/main/java/com/teamb/globalipbackend1/service/auth/RoleRequestService.java
package com.teamb.globalipbackend1.service.auth;


import java.util.List;
import java.util.Set;

import com.teamb.globalipbackend1.dto.authentication.RoleRequestAdminViewDto;
import com.teamb.globalipbackend1.model.user.Role;
import com.teamb.globalipbackend1.repository.user.RoleRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.teamb.globalipbackend1.model.authentication.RoleRequest;
import com.teamb.globalipbackend1.model.authentication.RoleRequestStatus;
import com.teamb.globalipbackend1.repository.user.RoleRequestRepository;
import com.teamb.globalipbackend1.model.user.User;
import com.teamb.globalipbackend1.repository.user.UserRepository;

@Service
public class RoleRequestService {

    private final RoleRequestRepository roleRequestRepo;
    private final UserRepository userRepo;
    private final RoleRepository roleRepository;

    public RoleRequestService(RoleRequestRepository roleRequestRepo,
                              UserRepository userRepo, RoleRepository roleRepository) {
        this.roleRequestRepo = roleRequestRepo;
        this.userRepo = userRepo;
        this.roleRepository = roleRepository;
    }

    public void requestAdminRole(String userId) {
        boolean alreadyPending =
                roleRequestRepo.existsByUserIdAndStatus(userId, RoleRequestStatus.PENDING);

        if (alreadyPending) {
            throw new IllegalStateException("Admin request already pending");
        }

        RoleRequest request = new RoleRequest(userId, "ROLE_ADMIN");
        roleRequestRepo.save(request);
    }


    public List<RoleRequestAdminViewDto> getPendingRequests() {

        List<RoleRequest> requests =
                roleRequestRepo.findByStatus(RoleRequestStatus.PENDING);

        return requests.stream()
                .map(req -> {
                    User user = userRepo.findById(req.getUserId())
                            .orElseThrow(() ->
                                    new RuntimeException("User not found for role request"));

                    return new RoleRequestAdminViewDto(
                            req.getId(),
                            user.getUserId(),
                            user.getUsername(),
                            user.getEmail(),
                            req.getRequestedRole(),
                            req.getRequestedAt()
                    );
                })
                .toList();
    }

    // ADMIN → Approve
    @Transactional
    public void approveRequest(String requestId, String adminId) {
        RoleRequest request = roleRequestRepo.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found"));

        User user = userRepo.findById(request.getUserId())
                .orElseThrow(() -> new RuntimeException("User not found"));

        Role adminRole = roleRepository.findByRoleType("ADMIN")
                .orElseGet(() -> {
                    Role newRole = new Role("ADMIN");
                    return roleRepository.save(newRole);
                });
        Set<Role> roles=user.getRoles();
        roles.add(adminRole);
        user.setRoles(roles);
        request.approve(adminId);
    }

    // ADMIN → Reject
    public void rejectRequest(String requestId, String adminId) {
        RoleRequest request = roleRequestRepo.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found"));

        request.reject(adminId);
        roleRequestRepo.save(request);
    }

    // ADMIN → Waitlist
    public void waitlistRequest(String requestId, String adminId) {
        RoleRequest request = roleRequestRepo.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found"));

        request.waitlist(adminId);
        roleRequestRepo.save(request);
    }
}

